[InstCombine] Fold `lshr -> zext -> shl` patterns

zGoldthorpe opened the pull request:
This patch to the InstCombiner eliminates one of the shift operations from `lshr -> zext -> shl` patterns such as

```llvm
define i64 @src_masked(i32 %x) {
  %lshr = lshr i32 %x, 4
  %mask = and i32 %lshr, 255
  %zext = zext i32 %mask to i64
  %shl = shl i64 %zext, 48
  ret i64 %shl
}
; =>
define i64 @tgt_masked(i32 %x) {
  %mask = and i32 %x, 4080
  %zext = zext nneg i32 %mask to i64
  %shl = shl nsw nuw i64 %zext, 44
  ret i64 %shl
}
```

In so doing, this enables the InstCombiner to fully simplify certain integer unpack/repack patterns such as

```llvm
define i64 @src_combine(i32 %lower, i32 %upper) {
  %base = zext i32 %lower to i64
  %u.0 = and i32 %upper, 255
  %z.0 = zext i32 %u.0 to i64
  %s.0 = shl i64 %z.0, 32
  %o.0 = or i64 %base, %s.0
  %r.1 = lshr i32 %upper, 8
  %u.1 = and i32 %r.1, 255
  %z.1 = zext i32 %u.1 to i64
  %s.1 = shl i64 %z.1, 40
  %o.1 = or i64 %o.0, %s.1
  %r.2 = lshr i32 %upper, 16
  %u.2 = and i32 %r.2, 255
  %z.2 = zext i32 %u.2 to i64
  %s.2 = shl i64 %z.2, 48
  %o.2 = or i64 %o.1, %s.2
  %r.3 = lshr i32 %upper, 24
  %u.3 = and i32 %r.3, 255
  %z.3 = zext i32 %u.3 to i64
  %s.3 = shl i64 %z.3, 56
  %o.3 = or i64 %o.2, %s.3
  ret i64 %o.3
}
; =>
define i64 @tgt_combine(i32 %lower, i32 %upper) {
  %base = zext i32 %lower to i64
  %1 = zext i32 %upper to i64
  %2 = shl nuw i64 %1, 32
  %o.3 = or disjoint i64 %2, %base
  ret i64 %o.3
}
```

Alive2 proofs: [T7U3aM](https://alive2.llvm.org/ce/z/T7U3aM)

zGoldthorpe committed:
Added patch to fold `lshr + zext + shl` patterns

zGoldthorpe commented:
@arsenm @jrbyrnes pinging to request review

zGoldthorpe committed:
Incorporated straightforward reviewer feedback.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -978,6 +978,47 @@ Instruction *InstCombinerImpl::foldLShrOverflowBit(BinaryOperator &I) {
+  // Dig through operations until the first shift.
+  while (!I->isShift())
+    if (!match(I, m_BinOp(m_OneUse(m_Instruction(I)), m_Constant())))
+      return nullptr;
```
An alternative is to fold two adjacent `unpack-pack` patterns into one in `InstCombinerImpl::foldDisjointOr`: https://alive2.llvm.org/ce/z/N25Hx_
```
define i64 @src(i32 %upper) {
  %r.1 = lshr i32 %upper, 8
  %u.1 = and i32 %r.1, 255
  %z.1 = zext nneg i32 %u.1 to i64
  %s.1 = shl nuw nsw i64 %z.1, 40

  %r.2 = lshr i32 %upper, 16
  %u.2 = and i32 %r.2, 255
  %z.2 = zext nneg i32 %u.2 to i64
  %s.2 = shl nuw nsw i64 %z.2, 48

  %o.1 = or disjoint i64 %s.2, %s.1
  ret i64 %o.1
}

define i64 @tgt(i32 %upper) {
  %r.1 = lshr i32 %upper, 8
  %u.1 = and i32 %r.1, 65535
  %z.1 = zext nneg i32 %u.1 to i64
  %s.1 = shl nuw nsw i64 %z.1, 40
  ret i64 %s.1
}
```
Then `InstCombinerImpl::reassociateDisjointOr` will handle other stuff.

zGoldthorpe committed:
Refactored `canEvaluateShifted` to identify candidates for
simplification.

zGoldthorpe commented:
@nikic @arsenm @dtcxzyw ping

zGoldthorpe commented:
@nikic @arsenm @dtcxzyw pinging again to review the refactoring of `canEvaluateShifted` to identify shift amounts that can be simplified.

ronlieb commented:
@shiltian   ok to approve and land ?

zGoldthorpe commented:
@nikic @arsenm @dtcxzyw should I request a review from someone else, or is there a deficiency in my current implementation?

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -978,6 +1025,50 @@ Instruction *InstCombinerImpl::foldLShrOverflowBit(BinaryOperator &I) {
+  // computation.
+  if (!refineEvaluableShiftMask(Op, ShrMask, /*IsLeftShift=*/true, IC, nullptr))
+    return nullptr;
+
```
I added an `if (!ShrMask.isPowerOf2()) return nullptr` check here, and all tests were still passed. I guess it may be helpful to avoid introducing new shifts in `getShiftedValue`. Not sure if it can avoid the regression in https://github.com/dtcxzyw/llvm-opt-benchmark/pull/2662/commits/876efdaecd0a0b9b8b03b3aa83ed1568274ea511#r2267345986.

arsenm commented on llvm/test/Transforms/InstCombine/shifts-around-zext.ll:
```diff
@@ -0,0 +1,162 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -S -passes=instcombine %s | FileCheck %s
+
+declare void @clobber.i32(i32)
```
s/clobber/use/?

zGoldthorpe committed:
clobber -> use

zGoldthorpe committed:
Removed dead code.

zGoldthorpe committed:
Made rule more conservative to keep IR more canonical.

Inner `lshr` instructions are only folded through the outer `shl` if its
immediate user is an `and` operation (since these shifts reduce to a
mask anyway).

zGoldthorpe committed:
Merge branch 'main' into pr/cancel-shift-thru-zext

dtcxzyw commented on llvm/test/Transforms/InstCombine/shifts-around-zext.ll:
```diff
@@ -128,9 +128,14 @@ define i64 @combine(i32 %lower, i32 %upper) {
+; CHECK-NEXT:    [[R_3:%.*]] = lshr i32 [[UPPER]], 24
+; CHECK-NEXT:    [[Z_3:%.*]] = zext nneg i32 [[R_3]] to i64
+; CHECK-NEXT:    [[S_3:%.*]] = shl nuw i64 [[Z_3]], 56
+; CHECK-NEXT:    [[O_3:%.*]] = or disjoint i64 [[O_2]], [[S_3]]
```
In my previous test https://github.com/llvm/llvm-project/pull/147737/files#r2267620032, bailing out on non-single-bit ShrMask should not affect the result.
I don't fully understand the meaning of `OpposingInnerShiftInfo`. The current implementation looks weird. `InnerShInfo` may be assigned multiple times. `ShrMask` is computed but not used.

zGoldthorpe replied:
The current implementation is a WIP, since the idea behind it is not at all satisfactory to me either.

The `OpposingInnerShiftInfo` is just a workaround for tracking the contextual operation of the inner shift so I can guard the fold to only occur when the inner shift happens inside of an `and`. While it's true it can be assigned to multiple times, I wouldn't say that's the biggest problem with the current commit.

I was hoping this would spark more conversation about what would be a more suitable strategy for making this pattern more conservative.

> In my previous test https://github.com/llvm/llvm-project/pull/147737/files#r2267620032, bailing out on non-single-bit ShrMask should not affect the result.

Sorry, I forgot to respond to that comment. I don't see how this condition is supposed to do anything. I tried your suggestion against the regressions, and they had no effect (i.e., the behaviour has identical regressions as the case where there is no early bail-out).

> `ShrMask` is computed but not used.

`ShrMask` is still used as a set of candidate shift amounts for `refineEvaluableShiftMask` to scan over, and allows for an early bail-out if the mask is zeroed out while recursing.
